### ISSUE TITLE
A single NaN batch disables training for the remainder of the epoch

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/autoslide#119]
Part of #116.

## The defect

https://github.com/pulakatlab/autoslide/blob/main/autoslide/src/pipeline/model/training_utils.py#L163-L172

```python
losses = sum([l for l in loss.values()])
train_epoch_loss += losses.cpu().detach().numpy()
if np.isnan(train_epoch_loss):
    print('Loss is NaN, skipping batch')
    continue
optimizer.zero_grad()
losses.backward()
optimizer.step()
```

The NaN check runs on the **accumulated epoch loss**, after the accumulation.
Once `train_epoch_loss` becomes NaN it stays NaN for the remainder of the epoch
(`nan + x == nan`), so every subsequent batch takes the `continue` branch and
skips `zero_grad`, `backward` and `step` entirely.

A single bad batch therefore silently disables training for the rest of that
epoch. The only visible symptom is a repeated log line and a flat loss curve.

## Suggested fix

Check the batch loss, not the running total, and guard the gradients:

```python
losses = sum(loss.values())
if not torch.isfinite(losses):
    print(f'Non-finite loss at batch {i}, skipping')
    optimizer.zero_grad(set_to_none=True)
    continue
train_epoch_loss += losses.item()
optimizer.zero_grad()
losses.backward()
torch.nn.utils.clip_grad_norm_(
    [p for p in model.parameters() if p.requires_grad], max_norm=10.0)
optimizer.step()
```

Gradient clipping is worth adding regardless — `lr=0.005` at batch size 2 with
no warmup is exactly the regime where detection losses spike early in training.

It would also be worth logging how often this branch fires in the existing
training runs, to find out whether any released checkpoint was affected.
